### PR TITLE
net-libs/telepathy-qt: add 0.9.8, bug #699882

### PR DESCRIPTION
**Pull request by Kaffeine**
The new version works with Python 3 and has fixed tests.

Signed-off-by: Alexander Akulich <akulichalexander@gmail.com>

Closes: https://bugs.gentoo.org/699882
Closes: https://bugs.gentoo.org/549448

### CODE REVIEW

File: net-libs/telepathy-qt/telepathy-qt-0.9.8.ebuild [hunk 1]
```diff
@@ -0,0 +1,62 @@
+# Copyright 1999-2019 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=7
+
+PYTHON_COMPAT=( python3_{5,6,7} )
+inherit python-single-r1 cmake-utils virtualx
```
Comment (a17r) [inline net-libs/telepathy-qt/telepathy-qt-0.9.8.ebuild]: python deps are apparently only added for BDEPEND and DEPEND (test), why the switch to python-single-r1?

File: net-libs/telepathy-qt/telepathy-qt-0.9.8.ebuild [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+DESCRIPTION="Qt bindings for the Telepathy D-Bus protocol"
+HOMEPAGE="https://telepathy.freedesktop.org/"
+SRC_URI="https://telepathy.freedesktop.org/releases/${PN}/${P}.tar.gz"
+
+LICENSE="LGPL-2.1"
+SLOT="0"
+KEYWORDS="~amd64 ~arm ~arm64 ~x86"
+IUSE="debug farstream test"
+REQUIRED_USE="${PYTHON_REQUIRED_USE}"
```
Comment (a17r) [inline net-libs/telepathy-qt/telepathy-qt-0.9.8.ebuild]: You had to add `REQUIRED_USE` for `python-single-r1`, but with `python-any-r1` it is wrong.

File: net-libs/telepathy-qt/telepathy-qt-0.9.8.ebuild [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	dev-qt/qtxml:5
+	farstream? (
+		>=net-libs/telepathy-farstream-0.2.2
+		>=net-libs/telepathy-glib-0.18.0
+	)
+	test? (
+		dev-libs/dbus-glib
+		dev-libs/glib:2
+		dev-qt/qttest:5
+		dev-python/dbus-python[${PYTHON_USEDEP}]
```
Comment (a17r) [inline net-libs/telepathy-qt/telepathy-qt-0.9.8.ebuild]: `${PYTHON_USEDEP}` exists in `python-single-r1`, but not in `python-any-r1`. Look here for what you need to do: https://wiki.gentoo.org/wiki/Project:Python/python-any-r1#Package_with_additional_dependencies

**Comment by Kaffeine** [comment]
@a17r, big thank you for the help! I hope I read the docs right :)

**Comment by Kaffeine** [comment]
Do we need to somehow block `<kde-apps/ktp-common-internals-19.08.3`?
On the other hand, it is OK to upgrade telepathy-qt with any Qt5-based `ktp-common-internals` installed, because the lib is 100% compatible on the client ABI level.

File: net-libs/telepathy-qt/telepathy-qt-0.9.8.ebuild (hunk 3 above)
Comment (Kaffeine) [reply net-libs/telepathy-qt/telepathy-qt-0.9.8.ebuild]: Meh, I copy-pasted the code from the link and didn't notice the unpaired parenthesis. @a17r would you mind to fix the example in wiki? I For some reason I'm unable to sign in there. :thinking:

**Comment by a17r** [comment]
> Do we need to somehow block `<kde-apps/ktp-common-internals-19.08.3`?
No need to, we will make sure to stabilise kde-apps-19.08.3 before or at the same time as telepathy-qt-0.9.8.

File: net-libs/telepathy-qt/telepathy-qt-0.9.8.ebuild (hunk 3 above)
Comment (a17r) [reply net-libs/telepathy-qt/telepathy-qt-0.9.8.ebuild]: thanks, it is fixed now ;)

File: net-libs/telepathy-qt/telepathy-qt-0.9.8.ebuild [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	dev-qt/qtcore:5
+	dev-qt/qtdbus:5
+	dev-qt/qtgui:5
+	dev-qt/qtnetwork:5
+	dev-qt/qtxml:5
+	farstream? (
+		>=net-libs/telepathy-farstream-0.2.2
+		>=net-libs/telepathy-glib-0.18.0
+	)
+	test? (
```
Comment (a17r) [inline net-libs/telepathy-qt/telepathy-qt-0.9.8.ebuild]: `test` depend in RDEPEND can not be right. This should all be in `DEPEND`.

File: net-libs/telepathy-qt/telepathy-qt-0.9.8.ebuild (hunk 2 above)
Comment (a17r) [reply net-libs/telepathy-qt/telepathy-qt-0.9.8.ebuild]: it seems you overlooked that comment - going to fix it for you together with the test dependencies on merge. ;)

**Comment by Kaffeine** [comment]
I'm sorry I had no time to fix this yesterday or on Friday, but I can try to fix it now.

OTOH it feels like I'm a dumb newbie :smile: burning CI CPU time for nothing. Maybe it's better indeed if I just give up and let you fix this for everyone.

File: net-libs/telepathy-qt/telepathy-qt-0.9.8.ebuild (hunk 4 above)
Comment (Kaffeine) [reply net-libs/telepathy-qt/telepathy-qt-0.9.8.ebuild]: I don't get the difference in `DEPEND` and `BDEPEND` (build dependencies). https://devmanual.gentoo.org/ebuild-writing/variables/ says to see the [Dependencies](https://devmanual.gentoo.org/general-concepts/dependencies/index.html) page, but the page has no info about `BDEPEND`.

**Comment by Kaffeine** [comment]
I pushed the new version with `test?` moved to `DEPEND`, but maybe it should be listed in `BDEPEND`.

File: net-libs/telepathy-qt/telepathy-qt-0.9.8.ebuild [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+	)
+"
+DEPEND="${RDEPEND}
+	test? (
+		dev-libs/dbus-glib
+		dev-libs/glib:2
+		dev-qt/qttest:5
+		$(python_gen_any_dep '
+			dev-python/dbus-python[${PYTHON_USEDEP}]
+		'
```
Comment (a17r) [inline net-libs/telepathy-qt/telepathy-qt-0.9.8.ebuild]: missing closing parenthesis reintroduced ;)

**Comment by a17r** [comment]
> I pushed the new version with removed `PYTHON_REQUIRED_USE` and `test?` moved to `DEPEND`, but maybe it should be listed in `BDEPEND`.

The best I can come up with is https://dev.gentoo.org/~mgorny/articles/the-ultimate-guide-to-eapi-7.html#build-time-dependencies-split-into-depend-and-bdepend

On a very basic level: if only binaries are used at build time, it should be in BDEPEND, as soon as it includes headers, then DEPEND.

**Comment by Kaffeine** [comment]
Mehh. It's the same typo as in b0f11ff7cb833bf10f5d49c51c580c2d31b95577 that I fixed in 94b714aa8b82cd6aba306a6ea2420c4cad93e1b3 on another PC. 

About the test: can you please run it with [TEST_VERBOSE=1](https://github.com/gentoo/gentoo/blob/master/eclass/cmake-utils.eclass#L784)? I've seen this before and I had the impression that the tests aren't running correctly.
I also set up CI for TelepathyQt and it's running well: https://travis-ci.com/TelepathyIM/telepathy-qt/builds/135973459#L2472

**Comment by a17r** [comment]
Here it is: https://gist.github.com/a17r/109662cdbb43f5470f4aa6cc7b5ddf26

**Comment by Kaffeine** [comment]
@a17r, I'm doing `ebuild telepathy-qt-0.9.8.ebuild test` inside a `while` (with `rm .tested` of course) in a script and it still succeeds for a hundred runs now.

The log says that a test can't register a service (which is nonsense).

**Comment by Kaffeine** [comment]
Ooooooh, I think I know!
`cmake-utils_src_test()` is running `ctest -j "$(makeopts_jobs)" --test-load "$(makeopts_loadavg)"`, but different tests are using the same DBus service names, so probably the next test fails because the current one still holding the name.
Can you please try to run the tests in a single thread?

BaseProtocol, BaseConnectionManager and many other tests register `org.freedesktop.Telepathy.ConnectionManager.cmtest` service to test against. It will take time to change that behavior to e.g. to give a unique part of the name, because we have this `cmtest` name in test data files and so on.

**Comment by a17r** [comment]
Only `BaseProtocol` failed now, amended above gist with the latest run. ;)

**Comment by Kaffeine** [comment]
> Only `BaseProtocol` failed now, amended above gist with the latest run. ;)

The current version runs `ctest -j 3 --test-load 999 --extra-verbose --output-on-failure --extra-verbose --output-on-failure`.
 I think that `ctest --output-on-failure ` (without multithreading) will pass.

**Comment by Kaffeine** [comment]
I just realized that I also have such an issue in another project. A network library has two dozens of client-server tests and all tests bind the same local addresses and ports. :man_facepalming:. It wasn't a problem because tests are passed in six to seven seconds and I had no idea (nor my IDE) to run them simultaneously.

Thank you for rising this up, although I have no clear solution yet.

**Pull request closed without merging**